### PR TITLE
[atom][atom-vllm] optimize MTP for GLM 5.2

### PR DESCRIPTION
**Pull request by whx-sjtu**
## Motivation

1. Optimize MTP performance
- Remove H2D
- Fuse MLA MTP decode metadata, position, and context-lens updates into a single Triton kernel.
- Add replicated vocab embedding support for GLM-5.2 target/MTP draft paths and include the env flag in the compile cache key.
- Optimize MTP draft token selection with distributed argmax instead of full-vocab logits gathering.
- Fuse two rms_norm + one cat into one fused kernel.

2. Enable MTP for vllm-plugin and optimize

### ATOM
```
local-completions ({'tokenizer': '/home/models/GLM-5.2-MXFP4', 'base_url': 'http://localhost:8004/v1/completions', 'num_concurrent': 64, 'max_retries': 3, 'tokenized_requests': False}), gen_kwargs: ({}), limit: None, num_fewshot: 20, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9227|±  |0.0074|
|     |       |strict-match    |    20|exact_match|↑  |0.9212|±  |0.0074|
```

### ATOM-vLLM
```
local-completions ({'tokenizer': '/home/models/GLM-5.2-MXFP4', 'base_url': 'http://localhost:8004/v1/completions', 'num_concurrent': 64, 'max_retries': 3, 'tokenized_requests': False}), gen_kwargs: ({}), limit: None, num_fewshot: 20, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9212|±  |0.0074|
|     |       |strict-match    |    20|exact_match|↑  |0.9219|±  |0.0074|
```

### CODE REVIEW

File: atom/model_ops/attention_mla.py [hunk 1]
```diff
@@ -870,6 +870,13 @@ def _forward_prefill_mla(
             paged_cu_seqlens_q = attn_metadata.sparse_cu_seqlens_q
             paged_kv_indptr = attn_metadata.sparse_kv_indptr
             paged_kv_indices = self.sparse_kv_indices_buffer
+            # Sparse attention treats each query token as its own length-1
+            # sequence, so kv_last_page_lens must have one entry per query token.
+            # attn_metadata.kv_last_page_lens is the DENSE per-seq buffer (length
+            # bs) on the has_cached prefix path; reading it here over-reads and
+            # causes an illegal memory access. Use the per-token sparse buffer,
+            # mirroring the decode sparse path.
+            kv_last_page_lens = attn_metadata.sparse_kv_last_page_lens
```
Comment (zejunchen-zejun) [inline atom/model_ops/attention_mla.py:875]: Hi, @gbyu-amd 
Can we simply here, as for now kv_last_page_lens represents dense MLA and sparse MLA at different time

Comment (gbyu-amd) [reply atom/model_ops/attention_mla.py:875]: refined it a bit

**Review by zejunchen-zejun — DISMISSED** [review]

File: recipes/atom_vllm/GLM-5.md [hunk 2]
```diff
@@ -57,6 +57,28 @@ vllm serve amd/GLM-5.2-MXFP4 \
...
+    --port 8004 \
+    --async-scheduling \
+    --load-format fastsafetensors \
+    --trust-remote-code \
+    --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE"}' \
+    --kv-cache-dtype fp8 \
+    --tensor-parallel-size 4 \
+    --gpu-memory-utilization 0.9 \
+    --max-num-batched-tokens 16384 \
+    --additional-config '{"online_quant_config": {"global_quant_config": "ptpc_fp8", "exclude_layer": ["lm_head", "model.embed_tokens", "*.mlp.gate", "model.layers.[0-9].mlp.*expert*", "model.layers.[1-6][0-9].mlp.*expert*", "model.layers.7[0-7].mlp.*expert*"]}}' \
```
Comment (junyyang-amd) [inline recipes/atom_vllm/GLM-5.md:76]: 模型已经是量化后的MXFP4，还需要走在线量化吗

**Review by valarLip — APPROVED** [review]

File: recipes/atom_vllm/GLM-5.md (hunk 2 above)
Comment (zejunchen-zejun) [reply recipes/atom_vllm/GLM-5.md:76]: 要的，这边量化的是attn附近的gemm

Comment (junyyang-amd) [reply recipes/atom_vllm/GLM-5.md:76]: Got it.

**Review by junyyang-amd — APPROVED** [review]

File: recipes/atom_vllm/GLM-5.md (hunk 2 above)
Comment (whx-sjtu) [reply recipes/atom_vllm/GLM-5.md:76]: 需要对主模型的attention部分以及草稿模型的全部做FP8量化

**Comment by zejunchen-zejun** [comment]
suggested to close and reopen to trigger the CI

**Review by zejunchen-zejun — APPROVED** [review]

**Comment by zejunchen-zejun** [comment]
pending on PR https://github.com/ROCm/aiter/pull/4227 to restore the CI

**Comment by zejunchen-zejun** [comment]
rebase to trigger the ci with new aiter

**Comment by zejunchen-zejun** [comment]
close to trigger rerun

**Comment by zejunchen-zejun** [comment]
will merge this PR as approved + most CI cases passed
M3 not pass as https://github.com/ROCm/ATOM/pull/1473 has not been merged
